### PR TITLE
noise2same + SwinIR

### PR DESCRIPTION
**Pull request by pchizhov**

### CODE REVIEW

File: noise2same/swinir.py [hunk 1]
```diff
@@ -0,0 +1,854 @@
...
+    """
+    Args:
+        x: (B, H, W, C)
+        window_size (int): window size
+    Returns:
+        windows: (num_windows*B, window_size, window_size, C)
+    """
+    B, H, W, C = x.shape
+    x = x.view(B, H // window_size, window_size, W // window_size, window_size, C)
+    windows = x.permute(0, 1, 3, 2, 4, 5).contiguous().view(-1, window_size, window_size, C)
```
Comment (papkov) [inline noise2same/swinir.py]: `x = einops.rearrange('b (h wh) (w ww) c -> (b h w) wh ww c', wh=window_size, ww=window_size)` looks identical

File: noise2same/swinir.py [hunk 2]
```diff
@@ -0,0 +1,854 @@
...
+    return windows
+
+
+def window_reverse(windows, window_size, H, W):
```
Comment (papkov) [inline noise2same/swinir.py]: single-capital-letter vars make me sad :(

File: noise2same/swinir.py [hunk 3]
```diff
@@ -0,0 +1,854 @@
...
+        windows: (num_windows*B, window_size, window_size, C)
+        window_size (int): Window size
+        H (int): Height of image
+        W (int): Width of image
+    Returns:
+        x: (B, H, W, C)
+    """
+    B = int(windows.shape[0] / (H * W / window_size / window_size))
+    x = windows.view(B, H // window_size, W // window_size, window_size, window_size, -1)
+    x = x.permute(0, 1, 3, 2, 4, 5).contiguous().view(B, H, W, -1)
```
Comment (papkov) [inline noise2same/swinir.py]: `x = einops.rearrange('(b h w) wh ww ... -> b (h wh) (w ww) ...', b=B, h=H, w=W)`

File: noise2same/swinir.py [hunk 4]
```diff
@@ -0,0 +1,854 @@
...
+        coords_h = torch.arange(self.window_size[0])
+        coords_w = torch.arange(self.window_size[1])
+        coords = torch.stack(torch.meshgrid([coords_h, coords_w]))  # 2, Wh, Ww
+        coords_flatten = torch.flatten(coords, 1)  # 2, Wh*Ww
+        relative_coords = coords_flatten[:, :, None] - coords_flatten[:, None, :]  # 2, Wh*Ww, Wh*Ww
+        relative_coords = relative_coords.permute(1, 2, 0).contiguous()  # Wh*Ww, Wh*Ww, 2
+        relative_coords[:, :, 0] += self.window_size[0] - 1  # shift to start from 0
+        relative_coords[:, :, 1] += self.window_size[1] - 1
+        relative_coords[:, :, 0] *= 2 * self.window_size[1] - 1
+        relative_position_index = relative_coords.sum(-1)  # Wh*Ww, Wh*Ww
```
Comment (papkov) [inline noise2same/swinir.py]: I would factor it out for tests (we might want to do the same for 3D later)

File: noise2same/swinir.py [hunk 5]
```diff
@@ -0,0 +1,854 @@
...
+        assert H % 2 == 0 and W % 2 == 0, f"x size ({H}*{W}) are not even."
+
+        x = x.view(B, H, W, C)
+
+        x0 = x[:, 0::2, 0::2, :]  # B H/2 W/2 C
+        x1 = x[:, 1::2, 0::2, :]  # B H/2 W/2 C
+        x2 = x[:, 0::2, 1::2, :]  # B H/2 W/2 C
+        x3 = x[:, 1::2, 1::2, :]  # B H/2 W/2 C
+        x = torch.cat([x0, x1, x2, x3], -1)  # B H/2 W/2 4*C
+        x = x.view(B, -1, 4 * C)  # B H/2*W/2 4*C
```
Comment (papkov) [inline noise2same/swinir.py]: this looks identical to https://pytorch.org/docs/stable/generated/torch.nn.PixelUnshuffle.html
or did I misunderstand the idea?

File: noise2same/swinir.py [hunk 6]
```diff
@@ -0,0 +1,854 @@
...
+        norm_layer (nn.Module, optional): Normalization layer. Default: None
+    """
+
+    def __init__(self, img_size=224, patch_size=4, in_chans=3, embed_dim=96, norm_layer=None):
+        super().__init__()
+        img_size = to_2tuple(img_size)
+        patch_size = to_2tuple(patch_size)
+        patches_resolution = [img_size[0] // patch_size[0], img_size[1] // patch_size[1]]
+        self.img_size = img_size
+        self.patch_size = patch_size
```
Comment (papkov) [inline noise2same/swinir.py]: not used below indeed. I'd consider removing it altogether

File: noise2same/swinir.py [hunk 7]
```diff
@@ -0,0 +1,854 @@
...
+            x = self.upsample(x)
+        elif self.upsampler == 'nearest+conv':
+            # for real-world SR
+            x = self.conv_first(x)
+            x = self.conv_after_body(self.forward_features(x)) + x
+            x = self.conv_before_upsample(x)
+            x = self.lrelu(self.conv_up1(torch.nn.functional.interpolate(x, scale_factor=2, mode='nearest')))
+            if self.upscale == 4:
+                x = self.lrelu(self.conv_up2(torch.nn.functional.interpolate(x, scale_factor=2, mode='nearest')))
+            x = self.conv_last(self.lrelu(self.conv_hr(x)))
```
Comment (papkov) [inline noise2same/swinir.py]: I'd drop SR parts from our code

File: noise2same/swinir.py [hunk 8]
```diff
@@ -0,0 +1,854 @@
...
+    width = (720 // upscale // window_size + 1) * window_size
+    model = SwinIR(upscale=2, img_size=(height, width),
+                   window_size=window_size, img_range=1., depths=[6, 6, 6, 6],
+                   embed_dim=60, num_heads=[6, 6, 6, 6], mlp_ratio=2, upsampler='pixelshuffledirect')
+    print(model)
+    print(height, width, model.flops() / 1e9)
+
+    x = torch.randn((1, 3, height, width))
+    x = model(x)
+    print(x.shape)
```
Comment (papkov) [inline noise2same/swinir.py]: move to tests or drop?

File: noise2same/dataset/getter.py [hunk 9]
```diff
@@ -14,20 +14,33 @@
 from noise2same.util import normalize_percentile
 
 
+def compute_pad_divisor(cfg: DictConfig):
+    if cfg.backbone_name == 'unet':
+        return 2 ** cfg.backbone.depth
+    elif cfg.backbone_name == 'swinir':
+        return cfg.backbone.patch_size * cfg.backbone.window_size
+    else:
+        return 1
```
Comment (papkov) [inline noise2same/dataset/getter.py]: `raise ValueError` here perhaps because we don't expect other backbones atm?

File: noise2same/dataset/getter.py [hunk 10]
```diff
@@ -14,20 +14,33 @@
 from noise2same.util import normalize_percentile
 
 
+def compute_pad_divisor(cfg: DictConfig):
```
Comment (papkov) [inline noise2same/dataset/getter.py]: could you please add a docstring?

File: noise2same/backbone/decode_head/decode_head.py [hunk 11]
```diff
@@ -0,0 +1,149 @@
+from abc import ABCMeta, abstractmethod
+
+import torch
+import torch.nn as nn
+from noise2same.backbone.ops import resize
+
+
+class BaseDecodeHead(nn.Module, metaclass=ABCMeta):
```
Comment (papkov) [inline noise2same/backbone/decode_head/decode_head.py]: if these file are copypaste it's best to reference the source (and check license)

File: noise2same/decode_head.py [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+                 out_channels: int,
+                 kernel_size: int,
+                 padding: int = 0,
+                 conv_bias: bool = False):
+        conv = nn.Conv2d(in_channels, out_channels, kernel_size, padding=padding, bias=conv_bias)
+        norm = nn.BatchNorm2d(out_channels)
+        super(ConvBlock, self).__init__(conv, norm)
+
+
+class BaseDecodeHead(nn.Module, metaclass=ABCMeta):
```
Comment (papkov) [inline noise2same/decode_head.py]: are decode_head.py, decode_head/decode_head.py, and decoder/decode_head.py identical (files nearby as well)?

File: noise2same/backbone/bsp_swinir.py [hunk 13]
```diff
@@ -158,7 +117,7 @@ def flops(self, N):
         return flops
 
 
-class BSPSwinTransformerBlock(nn.Module):
+class BSpSwinTransformerBlock(nn.Module):
     r""" Swin Transformer Block.
```
Comment (papkov) [inline noise2same/backbone/bsp_swinir.py:20]: please add a docstring about what this module is doing and how is it different from SwinTransformerBlock

File: noise2same/backbone/bsp_swinir.py [hunk 14]
```diff
@@ -242,32 +201,41 @@ def forward(self, x, x_size, mask=None):
...
 
         # W-MSA/SW-MSA (to be compatible for testing on images whose shapes are the multiple of window size
         if self.input_resolution == x_size:
             attn_mask = self.attn_mask
         else:
             attn_mask = self.calculate_mask(x_size).to(x.device)
+
+        # replicating the mask for the whole batch
+        if attn_mask is not None:
+            attn_mask = attn_mask.unsqueeze(0).repeat(B, 1, 1, 1)  # (B, nW, window_size, window_size)
```
Comment (papkov) [inline noise2same/backbone/bsp_swinir.py]: it would be ideal to have different masks per image. looking at [this line](https://github.com/pchizhov/noise2same.pytorch/blob/ef57faf7076eb320c5b5b00218c1ea2dc2b541bc/noise2same/backbone/bsp_swinir.py#L90), it should be technically possible. let's think about it

File: noise2same/backbone/bsp_swinir.py [hunk 15]
```diff
@@ -423,7 +343,7 @@ def flops(self):
         return flops
 
 
-class BSPRSTB(nn.Module):
+class BSpRSTB(nn.Module):
```
Comment (papkov) [inline noise2same/backbone/bsp_swinir.py]: would it be possible in this file to inherit from the respective swin_ir classes to avoid copy-paste?
